### PR TITLE
AB#967 Add a store interface

### PR DESCRIPTION
**Pull request by daniel-weisse**
This PR adds the store interface to the coordinator.
This allows different implementations and approaches of storing and persisting data used by the coordinator.

Currently only the standard implementation in `coordinator/core/stdcore.go` is used. This implementation behaves the same as the old implementation of the core.

The `Store` interface implements 5 methods:
* `Get`: sends a request in the form of a string to the store and receives a byte array and/or error as response
* `Put`: send a request to save a byte array in the store referenced by a string. This data can then be requested using `Get`.
* `SealState`: encrypts and persists the data of the store to disk
* `LoadState`: decrypts data saved by `SealState` and loads them into the store
* `SetEncryptionKey`: sets the encryption key for `LoadState` and `SealState`

Using the methods implemented in `storewrapper` the different methods of `Store` are abstracted into more usable ways by taking care of unit conversions etc. For example when requesting the activations of a specific marble we can just call `getActivations(marbleType)` and get the activations as a `uint`.

### CODE REVIEW

File: coordinator/core/storewrapper.go
Comment (m1ghtym0) [inline coordinator/core/storewrapper.go:30]: Should `Store` be in a separate module?
Not sure what's best practice here @thomasten?
Another thought: Should `Get` and `Put` have the prefix/family as an argument instead of joining them in the wrapper?
This would leave it to the plugins to handle the prefix/family. Some storage backend might support this for performance reasons.

Comment (daniel-weisse) [reply coordinator/core/storewrapper.go:30]: I think it is better to have `Get` and `Put` only take a single argument as the key. This leaves the interface as a simple key-value store.
I dont think the performance impact of joining or splitting a short string will be noticeable, even with thousands of requests per second.

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:149]: what if this returns another error?

File: coordinator/core/core.go
Comment (thomasten) [inline coordinator/core/core.go:172]: reduce duplication

**Review by thomasten — CHANGES_REQUESTED** [review]
I think this is the right direction. Next step should be adding transactions before adding new features.
@daniel-weisse @m1ghtym0 pls check if https://dev.azure.com/Edgeless/Edgeless/_workitems/edit/997 makes sense to do next.
I didn't do a complete review. Either @m1ghtym0 (if not already done) or @Nirusu should do this before we merge this.

File: coordinator/core/storewrapper.go
Comment (thomasten) [reply coordinator/core/storewrapper.go:30]: I agree with @m1ghtym0 regarding separate module.
@daniel-weisse It's not about string operation cost, but KV stores that support grouping keys. However, I think this is negligible for our use case. I don't have a strong opinion, only a slight preference to keep it simple.

File: coordinator/core/storewrapper.go
Comment (daniel-weisse) [reply coordinator/core/storewrapper.go:63]: No specific reason to convert to hex, other than maybe saving a very small amount of space.
We convert to string at this point to easily save the data as `[]byte`. Not choosing hex would result in strings with slightly more characters for large number of activations.

File: coordinator/core/storewrapper.go
Comment (m1ghtym0) [inline coordinator/core/storewrapper.go:171]:
```suggestion
	manifest, err := s.getManifest("main")
	if err != nil {
		return nil, err
	}
```

File: coordinator/core/core.go
Comment (m1ghtym0) [inline coordinator/core/core.go:45]: Why are we keeping `manifest` and `updateManifest` in the Core? They just keep duplicate values to the ones in the store, no?

**Review by m1ghtym0 — CHANGES_REQUESTED** [review]

File: coordinator/core/core.go
Comment (daniel-weisse) [reply coordinator/core/core.go:45]: Its more of a convenience thing than anything else. Since the manifest does not change after it has been set once, it is kept as a `manifest.Manifest` object in memory. This removes the need to load the manifest from the store every time we want to iterate over values set by the manifest (for example when generating secrets for a marble).

We can of course change this to instead fetch it just in time using the `storeWrapper` `getManfiest` method

Comment (m1ghtym0) [reply coordinator/core/core.go:45]: Yeah, I'd keep it simple and omit caching for now. Considering future changes and potentially a distributed Coordinator we should refrain from special cases at this point.

Comment (thomasten) [reply coordinator/core/core.go:45]: It may be best to parse the manifest once and insert the values into the store. But does not need to happen within this PR.

Comment (thomasten) [reply coordinator/core/core.go:45]: We can then also get rid of the update manifest.

**Review by m1ghtym0 — APPROVED** [review]

**Review by thomasten — APPROVED** [review]

**Pull request merged**
